File: 3kyu/alphabetic_anagrams.js
function fac(n) {
  let ans = 1;
  while (n > 0) ans *= n--;
  return ans;
}

function repeat(iter) {
  const dt = {};
  for (const a of iter) {
    if (dt[a] === undefined) {
      dt[a] = 0;
    }
    dt[a]++;
  }
  return Object.keys(dt).reduce((a, b) => a * fac(dt[b]), 1);
}

function listPosition(word) {
  let res = 1;
  const arr = word.split('').sort();
  for (const w of word) {
    const ndx = arr.indexOf(w);
    res += fac(arr.length) * ndx / repeat(arr) / arr.length;
    arr.splice(ndx, 1);
  }
  return res;
}

File: 3kyu/AlphabeticAnagrams.hs
module AlphabeticAnagrams where

import Data.List
import Data.Map.Strict qualified as M

lexiPos :: String -> Integer
lexiPos word = go word (sort word) 1
 where
  go [] _ res = res
  go (w : ws) arr res =
    let ndx = case elemIndex w arr of
          Just i -> i
          Nothing -> error "character not found"

        len = length arr

        contribution =
          fac (fromIntegral len)
            * fromIntegral ndx
            `div` repeat' arr
            `div` fromIntegral len

        arr' = removeAt ndx arr
     in go ws arr' (res + contribution)

fac :: Integer -> Integer
fac n = product [1 .. n]

repeat' :: String -> Integer
repeat' iter =
  product
    [ fac (fromIntegral n)
    | n <- M.elems counts
    ]
 where
  counts = M.fromListWith (+) [(c, 1 :: Int) | c <- iter]

removeAt :: Int -> [a] -> [a]
removeAt i xs = take i xs ++ drop (i + 1) xs
